src/Language/DifferentialDatalog/Compile.hs

```diff
@@ -524,13 +526,15 @@ compile d_unoptimized specname modules rs_code dir crate_types = do
     let dot_file = (?specname <.> "dot", depGraphToDot $ progDependencyGraph d_unoptimized)
     -- Apply optimizations; transform the program to a form that this
     -- compiler can handle; make sure the program has at least one relation.
-    let d = addDummyRel
-            $ simplifyDifferentiation
-            $ simplifyDelayedRels
-            $ optimize d_unoptimized
+    let d_optimized = addDummyRel
+                      $ simplifyDifferentiation
+                      $ simplifyDelayedRels
+                      $ optimize d_unoptimized
+    -- Compile away D3log location annotations.
+    let (d, d3log_rel_map) = processD3logAnnotations d_optimized
```

Comment:
If you do this earlier then later passes may not have to worry about strange relations with annotations anymore.

Reply:
The reason this is here is to help keep error messages clear.  This is true for many other bits of syntax that could be desugared early on in the compilation pipeline, but then the user would have to deal with error messages over the desugared code that they did not write.